interfaces/builtin/uio.go

```diff
@@ -0,0 +1,154 @@
+	"strings"
+
+	"github.com/snapcore/snapd/interfaces"
+	"github.com/snapcore/snapd/interfaces/apparmor"
+	"github.com/snapcore/snapd/interfaces/hotplug"
+	"github.com/snapcore/snapd/interfaces/udev"
+	"github.com/snapcore/snapd/snap"
+)
+
+const uioSummary = `allows access to specific uio device`
+
+const uioBaseDeclarationSlots = `
+  uio:
+    allow-installation:
+      slot-snap-type:
+        - core
+        - gadget
+    deny-auto-connection: true
+`
+
+type uioInterface struct{}
+
+func (iface *uioInterface) Name() string {
+	return "uio"
+}
+
+func (iface *uioInterface) StaticInfo() interfaces.StaticInfo {
+	return interfaces.StaticInfo{
+		Summary:              uioSummary,
+		BaseDeclarationSlots: uioBaseDeclarationSlots,
+	}
+}
+
+var uioPattern = regexp.MustCompile(`^/dev/uio[0-9]+$`)
+
+func (iface *uioInterface) path(slotRef *interfaces.SlotRef, attrs interfaces.Attrer) (string, error) {
+	var path string
+	if err := attrs.Attr("path", &path); err != nil || path == "" {
+		return "", fmt.Errorf("slot %q must have a path attribute", slotRef)
+	}
+	path = filepath.Clean(path)
+	if !uioPattern.MatchString(path) {
+		return "", fmt.Errorf("%q is not a valid UIO device", path)
+	}
+	return path, nil
+}
+
+func (iface *uioInterface) BeforePrepareSlot(slot *snap.SlotInfo) error {
+	_, err := iface.path(&interfaces.SlotRef{Snap: slot.Snap.InstanceName(), Name: slot.Name}, slot)
+	return err
+}
+
+func (iface *uioInterface) AppArmorConnectedPlug(spec *apparmor.Specification, plug *interfaces.ConnectedPlug, slot *interfaces.ConnectedSlot) error {
+	path, err := iface.path(slot.Ref(), slot)
+	if err != nil {
+		return nil
+	}
+	spec.AddSnippet(fmt.Sprintf("%s rwm,", path))
+	// Assuming sysfs_base is /sys/class/uio/uio[0-9]+ where the leaf directory
+	// name matches /dev/uio[0-9]+ device name, the following files exists or
+	// may exist:
+	//  - $sysfs_base/{name,version,event}
+	//  - $sysfs_base/maps/map[0-9]+/{addr,name,offset,size}
+	//  - $sysfs_base/portio/port[0-9]+/{name,start,size,porttype}
+	// The expression below matches them all as they all may be required for
+	// userspace drivers to operate.
+	spec.AddSnippet(fmt.Sprintf("/sys/devices/platform/**/uio/%s/** rw,", strings.TrimPrefix(path, "/dev/")))
+	return nil
+}
+
+func (iface *uioInterface) UDevConnectedPlug(spec *udev.Specification, plug *interfaces.ConnectedPlug, slot *interfaces.ConnectedSlot) error {
+	path, err := iface.path(slot.Ref(), slot)
+	if err != nil {
+		return nil
+	}
+	spec.TagDevice(fmt.Sprintf(`KERNEL=="%s"`, strings.TrimPrefix(path, "/dev/")))
+	return nil
+}
+
+func (iface *uioInterface) AutoConnect(*snap.PlugInfo, *snap.SlotInfo) bool {
+	// Allow what is allowed in the declarations
+	return true
+}
+
+func (iface *uioInterface) HotplugDeviceDetected(di *hotplug.HotplugDeviceInfo) (*hotplug.ProposedSlot, error) {
+	if di.Subsystem() != "uio" {
+		return nil, nil
+	}
+	slot := hotplug.ProposedSlot{
+		Name: strings.TrimPrefix(di.DeviceName(), "/dev/"),
```

Comment:
Because this makes it match /dev/uioN number. In the serial port case the number is not something you are attached to as much, as serial ports get different dynamic names due to USB-based serials dominating the test case. In this case it just makes it less surprising as to which UIO device you are using.

Reply:
Are UIO devices always going to come up the same? Eg, if I have two different cards with to different drivers, will they always come up the same? While I don't have experience with UIO, I would guess not based on the upstream documentation. One reason we have the named symlink in serial-port is so that userspace can use the predictable name. It seems that we would want the same thing here.... Forgive me if I missed something.

Reply:
I don't know - we only had one example that used device tree so it was fixed for as long as the device tree is fixed.

To answer your original question, we have no additional information, we only see the device name, hence the proposed slot uses device name as its name. In a device where more context is available statically (e.g. vendor with specific device tree and gadget) this can be overridden by statically defining a slot for the same device.